### PR TITLE
[analyzer][NFC] Turn NodeBuilderContext into a class

### PR DESCRIPTION
**Pull request by sona-tau**
From issue #73088. I changed `NodeBuilderContext` into a class. Additionally, there were some other mentions of the former being a struct which I also changed into a class. This is my first time working with an issue so I will be open to hearing any advice or changes that need to be done.

### CODE REVIEW

**Comment by steakhal** [comment]
Thanks for the PR.
I think it would make sense to not expose data members in a class' public api.
We might need some member functions along the way to make this possible.

**Review by steakhal — CHANGES_REQUESTED** [review]
See my previous reply.

**Comment by diego-est** [comment]
I added the appropriate functions and moved the class members into private. Additionally I ran `make check-all` to make sure the now-private members wouldn't cause any issues with the rest of the codebase, all the checks passed.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/CoreEngine.h [hunk 1]
```diff
@@ -194,11 +194,12 @@ class CoreEngine {
 };
 
 // TODO: Turn into a class.
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/CoreEngine.h:196]: Remove this TODO, as its completed.

**Review by steakhal — COMMENTED** [review]
Alright. So we can simply mark all the members private. I was surprised a bit.
LGTM, but remove the TODO comment you fix.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/CoreEngine.h (hunk 1 above)
Comment (diego-est) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/CoreEngine.h]: Done.

**Comment by diego-est** [comment]
Seems like it wasn't as easy as just making them private. I think I found all the places where there were private member accesses. The tests passed on my side again and the github-pull-requests action. The code_formatter action keeps failing because of the documentation on something I didn't touch (but seems to be fixed high upstream). Everything should be fixed now.

**Pull request merged**